app/controllers/categories_controller.rb

```diff
@@ -1,3 +1,81 @@
 class CategoriesController < ApplicationController
   
+  def index
+    @categories = Category.all
+  end
+  
+  def new
+    @categories  = Category.new
+  end
+  
+  def create
+    @category = Category.new(category_params) 
+    if @category.save 
+      flash[:success] = "Category added successfully"
+      redirect_to  root_path #merchant_path(@merchant)
+      return
+    else 
+      flash.now[:failure] = "Category failed to save"
+      render :new, status: :bad_request 
+      return
+    end
+  end
+  
+  # def show
+  #   category_id = params[:id]
+  #   @category = Category.find_by(id: category_id)
+    
+  #   if @category.nil?
+  #     head :not_found
+  #     return
+  #   end     
+  # end
+  
+  def edit
+    @category = Category.find_by(id: params[:id])
+    
+    if @category.nil?
+      head :not_found
+      return
+    end
+  end
+  
+  def update
+    @category = Category.find_by(id: params[:id])
+    
+    if @category.nil?
+      head :not_found
+      return
+    end
+
+    if @category.update(category_params)
+      redirect_to root_path 
+      return
+    else 
+      render :edit, status: :bad_request 
+      return
+    end
+  end
+  
+  def destroy
+    @category.destroy
+    
+    redirect_to categories_path
+    return
+  end
+  
+  private
+  
+  def category_params
+    return params.require(:category).permit(:name)
+  end
+  
+  def if_category_missing
+    if @category.nil?
+      flash[:error] = "category with id #{params[:id]} was not find"
+      redirect_to root_path
+      return
+    end
+  end
```

Comment:
I would suggest creating a private method called `find_category` that will run before the edit, update, and destroy actions.